backend-rust/src/indexer.rs

```diff
@@ -0,0 +1,779 @@
+            RewardsOverview::V1 {
+                common,
+                ..
+            } => common,
+        };
+        let total_amount = i64::try_from(common_reward_data.total_amount.micro_ccd())?;
+        let total_staked = match data.tokenomics_info {
+            RewardsOverview::V0 {
+                ..
+            } => {
+                // TODO Compute the total staked capital.
+                0i64
+            }
+            RewardsOverview::V1 {
+                total_staked_capital,
+                ..
+            } => i64::try_from(total_staked_capital.micro_ccd())?,
+        };
+
+        let mut prepared_block_items = Vec::new();
+        for block_item in data.events.iter() {
+            prepared_block_items.push(PreparedBlockItem::prepare(data, block_item)?)
+        }
+
+        Ok(Self {
+            hash,
+            height,
+            slot_time,
+            baker_id,
+            total_amount,
+            total_staked,
+            block_last_finalized,
+            prepared_block_items,
+        })
+    }
+
+    async fn save(
+        &self,
+        context: &mut BlockProcessor,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+    ) -> anyhow::Result<()> {
+        sqlx::query!(
+            r#"INSERT INTO blocks (height, hash, slot_time, block_time, baker_id, total_amount, total_staked)
+VALUES ($1, $2, $3,
+  (SELECT EXTRACT("MILLISECONDS" FROM $3 - b.slot_time) FROM blocks b WHERE b.height=($1 - 1::bigint)),
+  $4, $5, $6);"#,
+            self.height,
+            self.hash,
+            self.slot_time,
+            self.baker_id,
+            self.total_amount,
+            self.total_staked
+        )
+        .execute(tx.as_mut())
+            .await?;
+
+        // Check if this block knows of a new finalized block.
+        // If so, mark the blocks since last time as finalized by this block.
+        if self.block_last_finalized != context.last_finalized_hash {
+            let last_height = context.last_finalized_height;
+
+            let rec = sqlx::query!(
+                r#"
+WITH finalizer
+   AS (SELECT height FROM blocks WHERE hash = $1)
+UPDATE blocks b
+   SET finalization_time = EXTRACT("MILLISECONDS" FROM $3 - b.slot_time),
+       finalized_by = finalizer.height
+FROM finalizer
+WHERE $2 <= b.height AND b.height < finalizer.height
+RETURNING finalizer.height"#,
+                self.block_last_finalized,
+                last_height,
+                self.slot_time
+            )
+            .fetch_one(tx.as_mut())
+            .await
+            .context("Failed updating finalization_time")?;
+
+            context.last_finalized_height = rec.height;
+            context.last_finalized_hash = self.block_last_finalized.clone();
+        }
+
+        for item in self.prepared_block_items.iter() {
+            item.save(context, tx).await?;
+        }
+        Ok(())
+    }
+}
+
+struct PreparedBlockItem {
+    block_index:      i64,
+    tx_hash:          String,
```

Comment:
Since we have to convert it into a string before we can insert, I would like to do it as part of preprocessing.
If we are to make this more type safe then we should make newtype wrappers for every type, which could make sense as long as sqlx still can do compile-time validation of queries.